python/lsst/daf/butler/registries/sqlRegistry.py

```diff
@@ -482,17 +509,93 @@ def associate(self, collection, refs):
-                                 [{"dataset_id": ref.id, "collection": collection} for ref in refs])
+        # A collection cannot contain more than one Dataset with the same
+        # DatasetRef. Our SQL schema does not enforce this constaint yet so
+        # checks have to be done in the code:
+        # - read existing collection and try to match its contents with
+        #   new DatasetRefs using units
+        # - if there is a match and dataset_id is different then constraint
+        #   check fails
+        # TODO: This implementation has a race which can violate the
+        # constraint if multiple clients update registry concurrently. Proper
+        # constraint checks have to be implmented in schema.
+
+        def _matchRef(row, ref):
+            """Compare Dataset table row with a DatasetRef.
+
+            Parameters
+            ----------
+            row : `sqlalchemy.RowProxy`
+                Single row from Dataset table.
+            ref : `DatasetRef`
+
+            Returns
+            -------
+            match : `bool`
+                True if Dataset row is identical to ``ref`` (their IDs match),
+                False otherwise.
+
+            Raises
+            ------
+            ValueError
+                If DatasetRef unit values match row data but their IDs differ.
+            """
+            if row.dataset_id == ref.id:
+                return True
+
+            columns = self._dataUnits.getPrimaryKeyNames(ref.datasetType.dataUnits)
+            dataId = ref.dataId
+            if all(row[col] == dataId[col] for col in columns):
+                raise ValueError("A dataset of type {} with id: {} already exists in collection {}".format(
+                    ref.datasetType, dataId, collection))
+            return False
+
+        if len(refs) == 1:
+            # small optimization for a single ref
+            ref = refs[0]
+            dataset_id = self._findDatasetId(collection, ref.datasetType, ref.dataId)
+            if dataset_id == ref.id:
+                # already there
+                return
+            elif dataset_id is not None:
+                raise ValueError("A dataset of type {} with id: {} already exists in collection {}".format(
+                    ref.datasetType, ref.dataId, collection))
+        else:
+            # full scan of a collection to compare DatasetRef units
+            datasetTable = self._schema.tables["Dataset"]
+            datasetCollectionTable = self._schema.tables["DatasetCollection"]
+            query = datasetTable.select()
+            query = query.where(and_(datasetTable.c.dataset_id == datasetCollectionTable.c.dataset_id,
+                                     datasetCollectionTable.c.collection == collection))
+            result = self._connection.execute(query)
+            for row in result:
+                # skip DatasetRefs that are already there
+                refs = [ref for ref in refs if not _matchRef(row, ref)]
```

Comment:
This is a lot of list construction and re-construction.  Would it be better to pop `refs` from a `set` or `dict`?